src/ensembl/src/shared/components/upload/Upload.tsx

```diff
@@ -0,0 +1,187 @@
+  return fileReaderErrorMessages[error_code] || 'Unable to read the file.';
+};
+
+export enum FileReaderMethod {
+  ARRAY_BUFFER = 'readAsArrayBuffer',
+  BINARY_STRING = 'readAsBinaryString',
+  DATA_URL = 'readAsDataURL',
+  TEXT = 'readAsText'
+}
+type PropsForRespondingWithASingleFile = {
+  onChange: (file: File) => void;
+  callbackWithFiles: true;
+  allowMultiple: false;
+};
+
+type PropsForRespondingWithMultipleFiles = {
+  onChange: (files: FileList) => void;
+  callbackWithFiles: true;
+  allowMultiple: true;
+};
+
+type PropsForRespondingWithContent = {
+  onChange: (files: ReadFile[]) => void;
+  callbackWithFiles: false;
+  fileReaderMethod: FileReaderMethod;
+};
+
+type OnChangeProps =
+  | PropsForRespondingWithASingleFile
+  | PropsForRespondingWithMultipleFiles
+  | PropsForRespondingWithContent;
+
+export type UploadProps = {
+  id: string;
+  name?: string;
+  label: string;
+  classNames?: {
+    default?: string;
+    active?: string;
+  };
+  allowMultiple?: boolean;
+} & OnChangeProps;
+
+const Upload = (props: UploadProps) => {
+  const [drag, setDrag] = useState(false);
+
+  const handleDrag = (e: React.DragEvent) => {
+    e.preventDefault();
+    e.stopPropagation();
+  };
+  const handleDragIn = (e: React.DragEvent) => {
+    e.preventDefault();
+    e.stopPropagation();
+    if (get(e, 'dataTransfer.items.length', 0) > 0) {
+      setDrag(true);
+    }
+  };
+
+  const handleDragOut = (e: React.DragEvent) => {
+    e.preventDefault();
+    e.stopPropagation();
+    setDrag(false);
+  };
+
+  const handleSelectedFiles = async (
+    e: React.ChangeEvent<HTMLInputElement> | React.DragEvent
+  ) => {
+    e.preventDefault();
+    e.stopPropagation();
+    const eventType = e.type;
+    if (eventType === 'drop') {
+      (e as React.DragEvent).dataTransfer.clearData();
+    }
+    setDrag(false);
+
+    const files: FileList | null =
+      get(e, 'target.files') || get(e, 'dataTransfer.files') || null;
+
+    if (!files) {
+      return;
+    }
+
+    if (props.callbackWithFiles) {
+      // Just consider the first file if allowMultiple is true
+      if (!props.allowMultiple) {
+        props.onChange(files[0]);
+        return;
+      }
+      props.onChange(files);
+      return;
+    }
+
+    const filesToRead = props.allowMultiple
+      ? [...files]
+      : [...files].slice(0, 1);
+    const readFiles: ReadFile[] = [];
+
+    const promises = filesToRead.map((file) => {
+      const fileReader: FileReader = windowService.getFileReader();
+      return new Promise((resolve) => {
+        fileReader.onload = resolve;
+        fileReader.onerror = resolve;
+        fileReader[props.fileReaderMethod](file);
+      }).then(() => {
+        readFiles.push({
+          filename: file.name,
+          content: fileReader.result,
+          error: fileReader.error
+            ? getFileReaderErrorMessage(fileReader.error.code)
+            : null
+        });
+      });
+    });
```

Comment:
Still would like to push further against the mutable `readFiles` array, which now seems entirely unnecessary:

```
    const promises = filesToRead.map((file) => {
      const fileReader: FileReader = windowService.getFileReader();
      return new Promise((resolve) => {
        fileReader.onload = resolve;
        fileReader.onerror = resolve;
        fileReader[props.fileReaderMethod](file);
      }).then(() => ({
          filename: file.name,
          content: fileReader.result,
          error: fileReader.error
            ? getFileReaderErrorMessage(fileReader.error.code)
            : null
        }));
      });

    const results = await Promise.all(promises);
    props.onChange(readFiles));
```

poof, no `readFiles`!

Reply:
That's done